rbc/rbc.go

```diff
@@ -1,100 +1,483 @@
 }
 
 // handleMessage will distinguish input message (from ACS)
-func (rbc *RBC) handleMessage(sender cleisthenes.Member, msg *pb.Message) error {
-	panic("implement me w/ test case :-)")
+func (rbc *RBC) muxRequest(sender cleisthenes.Member, req cleisthenes.Request) error {
+	switch r := req.(type) {
+	case *ValRequest:
+		return rbc.handleValueRequest(sender, r)
+	case *EchoRequest:
+		return rbc.handleEchoRequest(sender, r)
+	case *ReadyRequest:
+		return rbc.handleReadyRequest(sender, r)
+	default:
+		return ErrInvalidRBCType
+	}
 }
 
-func (rbc *RBC) handleValueRequest(sender cleisthenes.Address, req ValRequest) error {
-	panic("implement me w/ test case :-)")
+func (rbc *RBC) handleValueRequest(sender cleisthenes.Member, req *ValRequest) error {
+	if rbc.valReceived {
+		return fmt.Errorf("already receive req message")
+	}
+
+	if rbc.echoSent {
+		return fmt.Errorf("already sent echo message - sender id : %s", sender.Address.String())
+	}
+
+	if !validateMessage(req) {
+		return fmt.Errorf("invalid VALUE request")
+	}
+
+	rbc.valReceived = true
+	rbc.echoSent = true
+	echoReq := &EchoRequest{*req}
+	rbc.shareMessage(rbc.proposer, echoReq)
+
+	iLogger.Infof(nil, "[VAL] onwer : %s, proposer : %s, sender : %s", rbc.owner.Address.String(), rbc.proposer.Address.String(), sender.Address.String())
+	return nil
 }
 
-func (rbc *RBC) handleEchoRequest(sender cleisthenes.Address, req EchoRequest) error {
-	panic("implement me w/ test case :-)")
+func (rbc *RBC) handleEchoRequest(sender cleisthenes.Member, req *EchoRequest) error {
+	if req, _ := rbc.echoReqRepo.Find(sender.Address); req != nil {
+		return fmt.Errorf("already received echo request - from : %s", sender.Address.String())
+	}
+
+	if !validateMessage(&req.ValRequest) {
+		return fmt.Errorf("invalid ECHO request - from : %s", sender.Address.String())
+	}
+
+	rbc.echoReqRepo.Save(sender.Address, req)
+
+	if rbc.countEchos(req.RootHash) >= rbc.echoThreshold() && !rbc.readySent {
+		rbc.readySent = true
+		readyReq := &ReadyRequest{req.RootHash}
+		rbc.shareMessage(rbc.proposer, readyReq)
+	}
+
+	if rbc.countReadys(req.RootHash) >= rbc.outputThreshold() && rbc.countEchos(req.RootHash) >= rbc.numDataShards {
+		value, err := rbc.interpolate(req.RootHash)
+		if err != nil {
+			return err
+		}
+		rbc.value = value
+	}
+
+	iLogger.Infof(nil, "[ECHO] owner : %s, proposer : %s, sender : %s", rbc.owner.Address.String(), rbc.proposer.Address.String(), sender.Address.String())
+	return nil
 }
 
-func (rbc *RBC) handleReadyRequest(sender cleisthenes.Address, req ReadyRequest) error {
-	panic("implement me w/ test case :-)")
+func (rbc *RBC) handleReadyRequest(sender cleisthenes.Member, req *ReadyRequest) error {
+	if req, _ := rbc.readyReqRepo.Find(sender.Address); req != nil {
+		return fmt.Errorf("already received ready request - from : %s", sender.Address.String())
+	}
+
+	rbc.readyReqRepo.Save(sender.Address, req)
+
+	if rbc.countReadys(req.RootHash) >= rbc.readyThreshold() && !rbc.readySent {
```

Comment:
1. how about refactor 289~304 logic into different function using channel?

Reply:
2. 289 if condition should be changed with 295 if condition

because in the case of `interpolate` failed we broadcast `Ready` request